File: Python/day15.py
#!/usr/bin/env python3
from typing import Tuple, Dict

from sortedcontainers import SortedList

from util import *

Coord = Tuple[int, int]
Map = Dict[Coord, Tuple[int, int]]  # position -> risk, total
Segment = [int, Coord]  # total risk, current node

DELTA = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def sorting_key(s: Segment) -> int:
    return s[0]


def find_path(risks: Map, start: Coord, end: Coord) -> int:
    visited = {start}
    queue: SortedList[Segment] = SortedList([(0, start)], key=sorting_key)
    while True:
        total_risk, (x, y) = queue.pop(0)
        if (x, y) == end:
            return total_risk
        for dx, dy in DELTA:
            new_pos = x + dx, y + dy
            if new_pos in risks and new_pos not in visited:
                risk, prev_total = risks[new_pos]
                new_risk = total_risk + risk
                if prev_total == -1 or new_risk < prev_total:
                    risks[new_pos] = risk, new_risk
                queue.add((new_risk, new_pos))
                visited.add(new_pos)


if __name__ == "__main__":
    data = load('day15.txt')
    risk_map = {(x, y): (int(h), -1)
                for y, line in enumerate(data)
                for x, h in enumerate(line)}

    best_path = find_path(risk_map, (0, 0), (99, 99))
    print('Lowest risk: %d' % best_path)
    assert best_path == 373

    risk_map = {
        (mx * len(line) + x, my * len(data) + y):
            ((int(h) - 1 + mx + my) % 9 + 1, -1)
        for my in range(5)
        for y, line in enumerate(data)
        for mx in range(5)
        for x, h in enumerate(line)}

    best_path = find_path(risk_map, (0, 0), (499, 499))
    print('Lowest risk: %d' % best_path)
    assert best_path == 2868

File: Elixir/lib/day15.ex
defmodule Day15 do
  @moduledoc """
  Documentation for `Day15`.
  """

  @delta [{-1, 0}, {1, 0}, {0, -1}, {0, 1}]
  @dim 100
  @multiply 5

  def input() do
    File.read!("data/day15.txt")
    |> String.split("\n")
    |> Stream.filter(&(String.length(&1) > 0))
    |> Stream.with_index()
    |> Enum.reduce(
         %{},
         fn {line, y}, acc ->
           String.split(line, "", trim: true)
           |> Stream.with_index()
           |> Enum.reduce(
                acc,
                fn {risk, x}, acc ->
                  Map.put(acc, {x, y}, String.to_integer(risk))
                end
              )
         end
       )
  end

  def add_sorted([], new_rp), do: [new_rp]
  def add_sorted([{risk, pos} | queue], {new_risk, new_pos}) when new_risk <= risk do
    [{new_risk, new_pos}, {risk, pos}] ++ queue
  end
  def add_sorted([rp | queue], new_rp), do: [rp | add_sorted(queue, new_rp)]

  def find_path([{total_risk, current} | _q], _v, _mr, _r, dest) when current == dest, do: total_risk
  def find_path([{total_risk, {x, y}} | queue], visited, max_risks, risks, dest) do
    {nq, nv, nmr} = @delta
                    |> Enum.reduce(
                         {queue, visited, max_risks},
                         fn {dx, dy}, {q, v, mr} ->
                           new_pos = {x + dx, y + dy}
                           if Map.has_key?(risks, new_pos) and !MapSet.member?(v, new_pos) do
                             new_risk = total_risk + risks[new_pos]
                             {
                               add_sorted(q, {new_risk, new_pos}),
                               MapSet.put(v, new_pos),
                               Map.update(mr, new_pos, new_risk, &(min(&1, new_risk)))
                             }
                           else
                             {q, v, mr}
                           end
                         end
                       )
    find_path(nq, nv, nmr, risks, dest)
  end
  def find_path(risks, start, dest) do
    visited = MapSet.new([start])
    queue = [{0, start}]
    find_path(queue, visited, %{}, risks, dest)
  end

  def find_path() do
    input()
    |> find_path({0, 0}, {@dim - 1, @dim - 1})
  end

  def  find_path_big() do
    input()
    |> Map.to_list()
    |> Enum.reduce(
         %{},
         fn {{x, y}, r}, acc ->
           Enum.to_list(0..@multiply - 1)
           |> Enum.reduce(
                acc,
                fn my, acc ->
                  Enum.to_list(0..@multiply - 1)
                  |> Enum.reduce(
                       acc,
                       fn mx, acc ->
                         Map.put(acc, {mx * @dim + x, my * @dim + y}, rem(r - 1 + mx + my, 9) + 1)
                       end
                     )
                end
              )
         end
       )
    |> find_path({0, 0}, {@multiply * @dim - 1, @multiply * @dim - 1})
  end
end
